apple-sdk: Include man pages starting from version 15

zhaofengli opened the pull request:
Apple SDKs include man pages for command-line utilities shipped with macOS as well as APIs, many of which are not available in source releases. It would be nice to have them in `apple-sdk`.

Since those man pages may conflict with those provided by Nixpkgs, we split them into a separate output named `sdkman` to avoid polluting devShells by default. In order to minimize rebuilds, the change is also gated to SDKs starting from version 15.

The man pages themselves appear to be compressible:

```
$ nix-store --dump /nix/store/26940l0019wnn7sg41x7zcz5ywimaxl3-apple-sdk-15.2-sdkman > apple-sdk-15.2-sdkman.nar
$ zstd apple-sdk-15.2-sdkman.nar -o apple-sdk-15.2-sdkman.nar.zst
apple-sdk-15.2-sdkman.nar : 58.24%   (  46.6 MiB =>   27.1 MiB, apple-sdk-15.2-sdkman.nar.zst)
```

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

zhaofengli commented on pkgs/by-name/ap/apple-sdk/common/split-man-pages-output.nix:
```diff
@@ -0,0 +1,35 @@
+}:
+
+self: super: let
+  # Tricky: fetchSDK cannot be called with boot fetchurl
+  isBoot = stdenv.name != "bootstrap-stage0-stdenv-darwin";
```
Not sure if this is the right solution. `fetchSDK` cannot be called with boot fetchurl. Normally this never happens, but if we try to evaluate it (triggered by `src.passthru.includeMan`) during bootstrap it fails:

```
       >        error: function 'anonymous lambda' called with unexpected argument 'version'
       >        at /nix/store/d2mfmgvy694mn1fy52wpzsr4lyksij4k-source/pkgs/build-support/fetchurl/boot.nix:7:1:
       >             6|
       >             7| {
       >              | ^
       >             8|   url ? builtins.head urls,
```

reckenrode replied:
That’s surprising because the Darwin stdenv bootstrap doesn’t use `fetchurl/boot.nix`. Does `--show-trace` have anything useful?

zhaofengli replied:
Sorry, missed this and just got home. This is surprising to me as well, and `--show-trace` is a whole mess. There are a few interesting lines which show it trying to evaluate the Linux stdenv (which somehow triggered apple-sdk):

```
       … while evaluating 'pkgs' to select 'formats.ini' on it
         at /Users/zhaofeng/Git/nixpkgs-2/nixos/modules/services/hardware/bluetooth.nix:26:12:
[...]
       … while evaluating 'prevStage' to select 'binutils-unwrapped' on it
         at /Users/zhaofeng/Git/nixpkgs-2/pkgs/stdenv/linux/default.nix:736:37:
```

Minimal diff to trigger this:

```diff
diff --git a/pkgs/by-name/ap/apple-sdk/package.nix b/pkgs/by-name/ap/apple-sdk/package.nix
index 04b32f07a642..1de364bb71ad 100644
--- a/pkgs/by-name/ap/apple-sdk/package.nix
+++ b/pkgs/by-name/ap/apple-sdk/package.nix
@@ -28,6 +28,7 @@ let
   sdkVersion = sdkInfo.version;
 
   fetchSDK = callPackage ./common/fetch-sdk.nix { };
+  src = fetchSDK sdkInfo;
 
   phases = lib.composeManyExtensions (
     [
@@ -39,7 +40,7 @@ let
       (callPackage ./common/process-stubs.nix { })
     ]
     # Avoid infinite recursions by not propagating certain packages, so they can themselves build with the SDK.
-    ++ lib.optionals (!enableBootstrap) [
+    ++ lib.optionals ((src ? whatever && false) || !enableBootstrap) [
       (callPackage ./common/propagate-inputs.nix { })
       (callPackage ./common/propagate-xcrun.nix { })
     ]
```

Run:

```
nix-instantiate --eval --strict --json --show-trace ./pkgs/top-level/release-attrpaths-superset.nix -A paths
```

zhaofengli replied:
Came back to this and went through the backtrace - Turns out it was caused by the fallback SDK conditional added in #346043. I opened #409574 to fix this.

reckenrode commented on pkgs/by-name/ap/apple-sdk/common/split-man-pages-output.nix:
```diff
@@ -0,0 +1,40 @@
+  passthru = super.src.passthru or { };
+  includeMan = passthru ? includeMan && passthru.includeMan;
+in
+if !isBoot && includeMan then
```
I apologize for not documenting this. There’s a blank `README.md` that I haven’t gotten back to writing yet. The SDK is built out of a series of overlays. The intent is they do what they do unconditionally with `package.nix` determining which overlays are applied. This condition should be moved there.

zhaofengli replied:
Ok, I moved all the conditionals to `package.nix`. Having to check for boot fetchurl is ugly, too. I moved the `lib.toInt (lib.versions.major sdkVersion) >= 15` logic to `package.nix` so we don't need to look into `src.passthru`.

zhaofengli committed:
apple-sdk: Include man pages in fetched SDK

We only include the man pages beginning with SDK version 15 to
minimize hash changes.

zhaofengli committed:
apple-sdk: Add phase to split man pages into a separate output

zhaofengli commented:
With #409574 merged, I reworked the PR to return to determining whether the fetched SDK has man pages with `src.passthru.includeMan`. This feels cleaner to me - `fetch-sdk` is responsible for pruning the raw SDK and can potentially be overridden, and it should therefore be the source of truth of whether the man pages are actually included.
